controller/lib/hal/hal_stm32.cpp

```diff
@@ -741,6 +743,28 @@ void HalApi::watchdog_handler() {
   wdog->key = 0xAAAA;
 }
 
+void HalApi::crc32_init() {
+  RCC_Regs *rcc = RCC_BASE;
+  // Enable clock to CRC32
+  rcc->periphClkEna[0] |= (1 << 12);
+  // Pull CRC32 peripheral out of reset if it ever was
+  rcc->periphReset[0] &= ~(1 << 12);
+
+  CRC_Regs *crc = CRC_BASE;
+  crc->init = 0xFFFFFFFF;
+  crc->poly = CRC32_POLYNOMIAL;
+  crc->ctrl = 1;
+}
+
+uint32_t HalApi::crc32(uint32_t d) {
+  CRC_Regs *crc = CRC_BASE;
+  crc->data = d;
+  // TODO CRC32 calculation takes 4 clock cycles,
+  // analyze ASM code to make sure we have enough commands to spend this time
+  // before reading CRC register
```

Comment:
Documentation is not entirely clear here. 

> 
> The CRC_DR register can be accessed by word, right-aligned half-word and right-aligned
> byte. For the other registers only 32-bit access is allowed.
> The duration of the computation depends on data width:
> • 4 AHB clock cycles for 32-bit
> • 2 AHB clock cycles for 16-bit
> • 1 AHB clock cycles for 8-bit
> An input buffer allows to immediately write a second data without waiting for any wait states
> due to the previous CRC calculation.

Apparently, if we want to just stream words into CRC, we can do that, but to read the crc result we have to wait 4 cycles.

I could rewrite this as
```
void crc32_accumulate(uint32_t d);
uint32_t get_crc32();
```
and maybe add 4 `asm("nops")` at the start of `get_crc32` to be more clear about this.

Reply:
I'm looking at https://www.st.com/resource/en/reference_manual/dm00151940-stm32l41xxx42xxx43xxx44xxx45xxx46xxx-advanced-armbased-32bit-mcus-stmicroelectronics.pdf, dunno if this is the right document.

14.2 says it has an "Input buffer to avoid bus stall during calculation".  I presume this means that it will stall under at least some circumstances.

Based on my naive reading, it sort of seems that it's saying you can write repeatedly to the register and it will compute the whole CRC, buffering the input and so on.  This should be faster than the one-in-one-out way we're currently doing it.

Could we find some example code that uses it?